### PR TITLE
 Added more filter coverage tests (list based) 

### PR DESCRIPTION
**Pull request by MacMcIrish**

### CODE REVIEW

File: tests/test_ModelSerialization.py [hunk 1]
```diff
@@ -187,3 +197,42 @@ def test_foreign_key_exposed(self, Classroom):
...
         assert len(classroom) == 1
         assert classroom[0]['school_id'] == self.school.id
+
+    def test_serialize_order_by_single_column(self, Student):
+        students = Student.serialize(
+            to_return=['name'],
+            order_by=Student.name
+        )
+        assert len(students) == 2
+        assert students[0]['name'] < students[1]['name']
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: <=

File: tests/test_ModelSerialization.py [hunk 2]
```diff
@@ -187,3 +197,42 @@ def test_foreign_key_exposed(self, Classroom):
...
+
+    def test_serialize_order_by_tuple(self, Student):
+        Student.serialize(
+            to_return=['name', 'email'],
+            order_by=(Student.name, Student.email)
+        )
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: where is the test?

File: tests/test_ModelFilter.py [hunk 3]
```diff
@@ -104,3 +104,27 @@ def test_filter_by_or_condition(self, Student):
...
+
+    def test_filter_by_list_multiple_values(self, Student):
+        assert len(Student.filter({
+            'id': [self.student1.id, self.student2.id]
+        }).all()) == 2
+
+    def test_filter_by_list_multiple_values_relationship(self, Teacher):
+        assert Teacher.filter({
+            'students.id': [self.student1.id, self.student2.id]
+        }).one().id == self.teacher.id
```
Comment (simlu) [inline tests/test_ModelFilter.py]: Doesn't really highlight the difference

File: tests/test_ModelSerialization.py [hunk 4]
```diff
@@ -187,3 +197,42 @@ def test_foreign_key_exposed(self, Classroom):
...
+
+    def test_serialize_order_by_tuple_with_id(self, Student):
+        Student.serialize(
+            to_return=['name', 'email'],
+            order_by=(Student.name, Student.email, Student.id)
+        )
+
+    def test_serialize_with_limit_and_offset(self, Student):
+        student_a = Student.serialize(
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: name result_a?

File: tests/test_ModelSerialization.py [hunk 5]
```diff
@@ -187,3 +197,42 @@ def test_foreign_key_exposed(self, Classroom):
...
+        assert student_a[0]['name'] in [self.student1.name, self.student2.name]
+        student_b = Student.serialize(
+            to_return=['name'],
+            limit=1,
+            offset=1,
+            order_by=Student.name
+        )
+        assert len(student_b) == 1
+        assert student_b[0]['name'] in [self.student1.name, self.student2.name]
+        assert student_a[0]['name'] != student_b[0]['name']
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: <=?

Comment (MacMcIrish) [reply tests/test_ModelSerialization.py]: It's not testing the order they're returned in, it's testing that the offsetting returns a different model for this test

File: tests/test_ModelSerialization.py [hunk 6]
```diff
@@ -187,3 +197,44 @@ def test_foreign_key_exposed(self, Classroom):
...
+        assert students[0]['name'] <= students[1]['name']
+
+    def test_serialize_order_by_tuple_with_id(self, Student):
+        students = Student.serialize(
+            to_return=['name', 'email'],
+            order_by=(Student.name, Student.email, Student.id)
+        )
+        assert students[0]['name'] <= students[1]['name']
+
+    def test_serialize_with_limit_and_offset(self, Student):
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: consider using teacher and return students.name - the interesting thing is that limit and offset work while fetching to many relationships (this is not trivial)

File: tests/test_ModelSerialization.py [hunk 7]
```diff
@@ -187,3 +197,44 @@ def test_foreign_key_exposed(self, Classroom):
...
+        assert students[0]['name'] <= students[1]['name']
+
+    def test_serialize_order_by_tuple(self, Student):
+        students = Student.serialize(
+            to_return=['name', 'email'],
+            order_by=(Student.name, Student.email)
+        )
+        assert students[0]['name'] <= students[1]['name']
+
+    def test_serialize_order_by_tuple_with_id(self, Student):
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: consider setting email and name the same

File: tests/test_ModelFilter.py [hunk 8]
```diff
@@ -104,3 +106,30 @@ def test_filter_by_or_condition(self, Student):
...
+            'id': [self.student1.id, self.student2.id]
+        }).all()) == 2
+
+    def test_filter_relationship_and_query(self, Teacher):
+        assert Teacher.filter({
+            'students.id': [self.student1.id, self.student2.id]
+        }).first() is not None
+        assert Teacher.filter({
+            'students.id': [self.student1.id, self.student3.id]
+        }).first() is None
```
Comment (simlu) [inline tests/test_ModelFilter.py]: add the "to one" relationship case as well

File: tests/test_ModelFilter.py [hunk 9]
```diff
@@ -104,3 +113,35 @@ def test_filter_by_or_condition(self, Student):
...
+
+    def test_filter_by_empty_list(self, Student):
+        assert Student.filter({'id': []}).first() is None
+
+    def test_filter_by_empty_list_relationship(self, Teacher):
```
Comment (simlu) [inline tests/test_ModelFilter.py]: test_filter_by_empty_list_relationship_to_one
Needs test_filter_by_empty_list_relationship_to_many?

File: tests/test_ModelSerialization.py [hunk 10]
```diff
@@ -187,3 +197,119 @@ def test_foreign_key_exposed(self, Classroom):
...
+        Student.filter({'id': self.student1.id}).one().update(
+            name=self.student1.name,
+            email=self.student1.email
+        ).save()
+        Student.filter({'id': self.student2.id}).one().update(
+            name=self.student2.name,
+            email=self.student2.email
+        ).save()
+
+    def test_serialize_order_by_tuple_with_id(self, Student):
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: test_serialize_order_by_tuple_with_id_as_fallback

File: tests/test_ModelSerialization.py [hunk 11]
```diff
@@ -187,3 +197,119 @@ def test_foreign_key_exposed(self, Classroom):
...
+            email='a@gmail.com'
+        ).save()
+        Student.filter({'id': self.student2.id}).one().update(
+            name='a',
+            email='a@gmail.com'
+        ).save()
+
+        students = Student.serialize(
+            to_return=['id', 'name', 'email'],
+            order_by=(Student.name, Student.email, Student.id),
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: No need to explicit order by id here

File: tests/test_ModelSerialization.py [hunk 12]
```diff
@@ -187,3 +197,119 @@ def test_foreign_key_exposed(self, Classroom):
...
+
+        # revert
+        Student.filter({'id': self.student1.id}).one().update(
+            name=self.student1.name
+        ).save()
+        Student.filter({'id': self.student2.id}).one().update(
+            name=self.student2.name
+        ).save()
+
+    def test_serialize_with_limit_and_offset_returns_to_many(self, Teacher):
```
Comment (simlu) [inline tests/test_ModelSerialization.py]: Let's remove this one - we'll deal with that in the sql tests

File: tests/test_ModelFilter.py [hunk 13]
```diff
@@ -104,3 +113,41 @@ def test_filter_by_or_condition(self, Student):
         assert len(students) == 2
         assert self.get(students, 'id', self.student1.id) is not None
         assert self.get(students, 'id', self.student2.id) is not None
+
+    def test_filter_by_list_single_value(self, Student):
```
Comment (simlu) [inline tests/test_ModelFilter.py]: remove by

**Pull request merged**